File: web/boot.js
// Starts CRuby in the page and hands it Hackety Hack.
import { RubyVM } from "@ruby/wasm-wasi";
import { WASI, OpenFile, File, Directory, PreopenDirectory, ConsoleStdout } from "@bjorn3/browser_wasi_shim";
import { unpack, buildTree } from "./vfs.mjs";

const params = new URLSearchParams(location.search);

async function boot() {
  const [wasmModule, vfsBuffer] = await Promise.all([
    WebAssembly.compileStreaming(fetch("/ruby.wasm")),
    fetch("/dist/vfs.bin").then((r) => r.arrayBuffer()),
  ]);

  const { root, dirFor } = buildTree(unpack(vfsBuffer), { File, Directory });
  // Two directories nothing ships files into but Ruby and the app both write
  // to: HOME is where Hackety Hack keeps ~/.hacketyhack.
  dirFor(["home"]);
  dirFor(["tmp"]);

  const env = [
    "HOME=/home",
    "TMPDIR=/tmp",
    "RUBYOPT=-EUTF-8",
    `CLOGS_ENTRY=${params.get("entry") || "/hh/hacketyhack.rb"}`,
  ];

  // `?env.FOO=bar` sets an environment variable, which is how the Shoes
  // programs that read one -- the benchmarks in tools/, say -- are configured
  // from a URL.
  for (const [key, value] of params) {
    if (key.startsWith("env.")) env.push(`${key.slice(4)}=${value}`);
  }

  const fds = [
    new OpenFile(new File([])),
    ConsoleStdout.lineBuffered((line) => console.log(line)),
    ConsoleStdout.lineBuffered((line) => console.warn(line)),
    new PreopenDirectory("/", root),
  ];

  const wasi = new WASI(["ruby", "/boot.rb"], env, fds, { debug: false });
  const { vm } = await RubyVM.instantiateModule({ module: wasmModule, wasip1: wasi });

  window.__clogsVM = vm;
  vm.eval(`load "/boot.rb"`);
  window.__clogsReady = true;
}

window.__clogsStatus = "booting";
window.__clogsReady = false;
boot().catch((error) => {
  window.__clogsStatus = "error";
  window.__clogsError = String(error && error.stack ? error.stack : error);
  console.error(error);
});

File: web/boot.rb
# frozen_string_literal: true

# The first Ruby the browser runs.
#
# Everything here is about making a wasm process look enough like a normal one
# for an unmodified Hackety Hack to boot in it: a load path, a working
# directory, a home directory to write into, and the two C extensions the app
# requires replaced by shims (see web/shims).

$LOAD_PATH.unshift("/shims", "/gems/lacci", "/gems/scarpe-components", "/gems/chunky_png", "/clogs", "/hh")

ENV["CLOGS_BACKEND"] = "wasm"

require "js"

module ClogsBoot
  module_function

  def report(status, message = nil)
    JS.global[:window][:__clogsStatus] = status
    JS.global[:window][:__clogsError] = message if message
    JS.global[:console].call(status == "error" ? :error : :log, "[clogs] #{status}#{message ? ": #{message}" : ""}")
  end

  def run
    # Hackety Hack takes HH::HOME from Dir.pwd, and its own directory is what
    # that has to be.
    Dir.chdir("/hh")

    require "clogs"

    entry = ENV["CLOGS_ENTRY"].to_s
    entry = "/hh/hacketyhack.rb" if entry.empty?

    # Hackety Hack's own entry point loads the Shoes 3 compatibility layer
    # itself. A bundled sample does not -- they are Shoes 3-era programs and
    # need it to run at all -- so put it in front of them, which is what
    # `rake samples` and the `clogs` executable both do.
    require "lib/compat/shoes3" unless entry.end_with?("hacketyhack.rb", "h-ety-h.rb")

    report("loading")
    load entry
    report("ready")
  rescue Exception => e
    report("error", "#{e.class}: #{e.message}")
    warn "#{e.class}: #{e.message}"
    warn e.backtrace.first(30).join("\n") if e.backtrace
  end
end

ClogsBoot.run
